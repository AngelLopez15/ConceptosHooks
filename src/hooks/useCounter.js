import {useState} from 'react'

// el inicial esta le ponemos que por default sea cero en caso de que no manden nada
export const useCounter = (initialState=0) => {
    const [state, setState] = useState(initialState)

    // factor es el parametro que vamos a recibir para ver de cuanto en cuanto va a ser
    // el incremento
    const incremento = (factor) => {
        setState(state+factor)
    }

    const decremento = (factor) =>{
        setState(state-factor)
    }

    // funcion para volver al valor por default
    const reset= () =>{
        setState(initialState)
    }
    
    // retornamos un objeto con las funciones y el estado
    return{
        state,
        incremento,
        decremento,
        reset
    }

}
