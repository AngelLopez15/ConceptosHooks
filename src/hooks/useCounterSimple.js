import {useState} from 'react'

// el inicial esta le ponemos que por default sea cero en caso de que no manden nada
export const useCounterSimple = (initialState=1) => {
    const [counter, setCounter] = useState(initialState)

    // factor es el parametro que vamos a recibir para ver de cuanto en cuanto va a ser
    // el incremento
    const incremento = () => {
        setCounter(counter+1)
    }

    const decremento = () =>{
        setCounter(counter-1)
    }

    // funcion para volver al valor por default
    const reset= () =>{
        setCounter(initialState)
    }
    
    // retornamos un objeto con las funciones y el estado
    return{
        counter,
        incremento,
        decremento,
        reset
    }

}
