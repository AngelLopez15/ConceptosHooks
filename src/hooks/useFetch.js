import { useState, useEffect, useRef } from "react"


export const useFetch = (url) => {

    // Con useRef vamos a manejar el estado de nuestro componenete. Va a iniciar en true
    // Por que cuando se inicia la aplicacion queremos ver el componenete. Sin embargo
    // cuando se hace la peticion queremos que ese componenete se oculte o que no haga
    // nada aunque le den clock y esto es para evitar
    // que lo esten presionando y provoque que nuestra aplicacion crashee
    const isMounted = useRef(true)

    const [state, setState] = useState({data:null, loading:true, error:null})

    // para manejar el useRef vamos a combinarlo con un useEffect que no hace nada mas
    // que pasar el valor de is mounted a false cuando se dispare el evento
    useEffect(()=>{
        return () =>{
            isMounted.current=false
        }
        // la dependencia vacia para que solo lo haga cunado el componenete se carga por primera vez
    },[])

    useEffect(()=>{

        // volviendo a poner los valores por default al hacer una peticion
        // para que el loading aparezca entre peticiones de frese y frese
        setState({data:null, loading:true, error:null})

        fetch(url)
            .then(resp=>resp.json())
            .then(data=>{

                // varificamos el estado del isMounted para manejar el estado de la 
                // peticion
                if (isMounted.current) {
                    setState({
                        loading:false,
                        error:null,
                        data
                    })
                }
            })
            .catch(()=>{
                setState({
                    data:null,
                    loading:false,
                    error:'No se pudo cargar la informaci√≥n'
                })
            })

    },[url])

    return state
}
