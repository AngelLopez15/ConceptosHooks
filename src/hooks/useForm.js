import {useState} from 'react'

export const useForm = (initialState={}) => {
    const [values, setValues] = useState(initialState)

    // NOTA DE INGLES: handle significa encargarse de
    
    // del evento "e" estamos desestructurando el target
    const handleInputChange = ({target}) => {
        setValues({
            ...values,
            [target.name]:target.value
        })
    }

    // podemos retornarlo como queramos en arreglo o en objeto
    return [values, handleInputChange]
}
