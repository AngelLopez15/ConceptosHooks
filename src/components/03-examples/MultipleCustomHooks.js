import React from 'react'

import '../02-useEffect/effect.css'
import { useFetch } from '../../hooks/useFetch'
import {useCounterSimple} from '../../hooks/useCounterSimple'

export const MultipleCustomHooks = () => {

    // importando los unicos propiedades que me interesan
    const {counter, incremento} = useCounterSimple(1)
    
    const {loading, data}= useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`)

    // extrayendo el autor y la cita de la data que manda la API
    // Una forma rapida de validar que venga la data es con 
    // una expresion logica. !!null <- estamos conviertiendo un null en un valor
    // booleano !null = true !!null = false
    // entonces estamos diciendo Si existe la data entonces que extraiga 
    // la posicion cero de la data. Si no trae nada la data es null por lo tanto
    // da false y ya no se ejecuta nada. PEro si la data trae algo va a ser true
    // y extraera la data
    const {author, quote} = !!data && data[0]  

    return (
        <div>
            <h1>BreakingBad Quotes</h1>
            <hr />
            {/* como tenemos dos componenetes uno se debe mostar cuando esta cargando
            y el otro cuando la API responde entonces usaremos un ternario */}
            {
                // si loadingexiste cargar el componente
                loading
                ?
                (
                <div className="alert alert-info text-center">
                    loading...
                </div>
                )
                // si no existe loading cargar el otro componente
                :
                (
                <blockquote className="blockquote text-right"> 
                    <p className="mb-0">{quote}</p>
                <footer className="blockquote-footer">{author}</footer>
                </blockquote>
                )
            }
            <button className="btn btn-primary"
                onClick={incremento}
            >
                Siguiente frase
            </button>
        </div>
    )
}
