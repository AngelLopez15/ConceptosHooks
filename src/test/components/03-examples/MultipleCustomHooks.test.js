import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks"
import toJson from 'enzyme-to-json';
import { useFetch } from '../../../hooks/useFetch';
import { useCounterSimple } from '../../../hooks/useCounterSimple';

jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounterSimple')

describe('Prueebas en <MultipleCustomHooks />', () => {
  configure({adapter: new Adapter()});

  useCounterSimple.mockReturnValue({
    counter:1,
    incremento:()=>{}
  })

  test('Debe de mostarse correctamente', () => {
    
    useFetch.mockReturnValue({
      data:null,
      loading:null,
      error:null
    })

    const wrapper = shallow(<MultipleCustomHooks /> )

    expect(toJson(wrapper)).toMatchSnapshot();
  })
  
  test('Debe de mostrar la informaciÃ³n', () => {

    useFetch.mockReturnValue({
      data:[{
        author:'Angel',
        quote:'Hola, Mundo'
      }],
      loading:false,
      error:null
    })

    const wrapper = shallow(<MultipleCustomHooks /> )
    expect(wrapper.find('.alert').exists()).toBe(false)
    expect(wrapper.find('.mb-0').text().trim()).toBe('Hola, Mundo')
    expect(wrapper.find('footer').text().trim()).toBe('Angel')

  });


})
