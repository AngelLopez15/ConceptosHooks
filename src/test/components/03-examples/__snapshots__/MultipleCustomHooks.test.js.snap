// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueebas en <MultipleCustomHooks /> Debe de mostarse correctamente 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-0"
    />
    <footer
      className="blockquote-footer"
    />
  </blockquote>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente frase
  </button>
</div>
`;
